### PR TITLE
[codex] Deepen Snyk source runtime

### PR DESCRIPTION
**Pull request by jonathanhaaswriter**
## Summary
- adds deeper Snyk REST runtime families for access, audit, cloud, collections, asset relationships, issues, and vulnerabilities
- moves Snyk family definitions into an internal runtime package and keeps Source CDK catalog/deploy metadata aligned
- projects Snyk memberships, service accounts, audit records, cloud resources, and asset relationships into graph evidence
- adds fixture coverage and nested JSON API list-key support for provider response shapes

## Validation
- go test ./sources/internal/jsonapi ./sources/snyk ./internal/sourceprojection -run 'Test(ReadUsesNestedConfiguredListKeys|Snyk|RuntimeUsesSnyk|NewFixture)' -count=1\n- make catalog-check\n- make sourcegen-check\n- make sourcegen-test\n- make check-arch\n- make connector-catalog-maintenance

---

<a href="https://app.devin.ai/review/writer/cerebro/pull/1643" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

File: sources/internal/snykapi/snykapi.go [hunk 1]
```diff
@@ -0,0 +1,653 @@
...
+		IDKeys:        []string{"id"},
+		ListKeys:      []string{"data"},
+		TimestampKeys: []string{"attributes.finished_at", "attributes.created_at"},
+		Attributes: map[string]string{
+			"org_id":          orgIDConfig,
+			"scan_id":         "id",
+			"source_event_id": "id",
+			"resource_id":     "id",
+			"resource_name":   "attributes.name|id",
+			"resource_type":   "scan",
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sources/internal/snykapi/snykapi.go]: Fixed in ef0b71e77. Removed the dead dynamic `resource_type: "scan"` mapping from the cloud scans family; cloud scan events now rely on the existing static `resource_type = "snyk_cloud_scan"`. Added an explicit runtime assertion for that attribute in `TestRuntimeUsesSnykRESTPathsAndVersionedPagination`. Verified with `go test ./sources/snyk ./sources/internal/snykapi ./sources/internal/jsonapi ./internal/sourceprojection -count=1` and `make connector-contract-check detection-catalog-check`.

File: sources/internal/snykapi/snykapi.go [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+	FamilyCloudScans       = "cloud_scans"
+	FamilyGroupMemberships = "group_memberships"
+	FamilyGroupSvcAccounts = "group_service_accounts"
+	FamilyGroupAuditLogs   = "group_audit_logs"
+	FamilyAssetProjects    = "asset_project_relationships"
+	FamilyAssetTargets     = "asset_target_relationships"
+)
+
+const (
+	apiVersionConfig       = APIVersionConfig
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sources/internal/snykapi/snykapi.go]: Fixed in 51f767f6e: removed the unused apiVersionConfig alias. Verified with `go test ./sources/snyk ./sources/internal/snykapi ./sources/internal/jsonapi ./internal/sourceprojection -count=1` and `make connector-contract-check detection-catalog-check`.

[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sources/internal/snykapi/snykapi.go]: Correction on the SHA in my previous reply: the pushed fix commit is 93771d677.

File: internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml [hunk 3]
```diff
@@ -341,25 +319,239 @@ entries:
...
+          label: Access review
+        - default_enabled: true
+          families: [audit_logs]
+          id: audit-activity
+          label: Audit activity
+        - default_enabled: true
+          families: [projects, targets, collections]
+          id: application-inventory
+          label: Application inventory
+        - default_enabled: true
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml:866]: Fixed in 67c13a69b. The connector catalog now exposes the org-scoped `cloud_environments` and `cloud_scans` families alongside `cloud_resources`, and the cloud-inventory scope includes all three cloud families plus assets. Because the connector catalog proof gate caps exposed resource families at 12, I kept the direct source runtime support for collections/package vulnerabilities but moved the catalog surface toward access, audit, findings, app inventory, and complete cloud inventory. Verified with `go test ./sources/snyk ./sources/internal/snykapi ./sources/internal/jsonapi ./internal/connectorcatalog -count=1` and `make connector-contract-check detection-catalog-check`.

File: sources/internal/snykapi/snykapi.go [hunk 4]
```diff
@@ -0,0 +1,651 @@
...
+			"created_at":      "attributes.created_at",
+			"resource_id":     "relationships.user.data.id",
+			"resource_type":   "relationships.user.data.type",
+		},
+		StaticAttributes: snykStaticAttributes("org_memberships", "identity_group_membership", "snyk_org_membership"),
+	})
+	family.Config.IdentityKeys = []string{orgIDConfig, "relationships.user.data.id"}
+	family.Config.ResourceURNKind = "runtime_users"
+	return family
+}
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sources/internal/snykapi/snykapi.go:292]: Fixed in 5cc1a3888: org_memberships now carries group_id from the org_id config attribute, and the live REST-path test asserts group_id=org-1 for membership payloads that do not include org_id in the record body. Verified with go test ./sources/snyk ./sources/internal/snykapi ./sources/internal/jsonapi ./internal/sourceprojection -count=1 and make catalog-check detection-catalog-check connector-contract-check.

File: internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml [hunk 5]
```diff
@@ -295,71 +276,280 @@ entries:
...
+          families: [assets, cloud_environments, cloud_resources, cloud_scans]
+          id: cloud-inventory
+          label: Cloud inventory
         - default_enabled: true
-          families:
-            - findings
-            - vulnerabilities
+          families: [findings]
           id: remediation
           label: Remediation
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml]: Resolved as intentional for this connector-catalog definition. The catalog proof gate allows at most 12 resource families; keeping a separate vulnerabilities catalog family would force out one of the cloud inventory families added here. The Snyk runtime, source catalog, deploy config, fixture suite, and projection registry continue to support snyk.vulnerabilities directly with the package_vulnerability query, while the connector catalog exposes remediation through findings.

File: internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml [hunk 6]
```diff
@@ -295,71 +276,280 @@ entries:
...
+          label: Application inventory
+        - default_enabled: true
+          families: [assets, cloud_environments, cloud_resources, cloud_scans]
+          id: cloud-inventory
+          label: Cloud inventory
         - default_enabled: true
-          families:
-            - findings
-            - vulnerabilities
+          families: [findings]
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml]: Fixed in 0c3f11015. The built-in Snyk connector catalog restores vulnerabilities to remediation and includes findings plus vulnerabilities in that scope; cloud_scans was removed from the built-in connector surface to stay within the 12-family proof gate while remaining supported by the source runtime/catalog.

File: internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml [hunk 7]
```diff
@@ -295,71 +276,280 @@ entries:
...
+          id: access-review
+          label: Access review
+        - default_enabled: true
+          families: [audit_logs]
+          id: audit-activity
+          label: Audit activity
+        - default_enabled: true
+          families: [projects, targets]
+          id: application-inventory
+          label: Application inventory
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml:865]: Resolved in 0c3f11015. Collections remain supported by the source runtime/source catalog, but the built-in connector catalog is capped at 12 resource families; this change prioritizes restoring remediation vulnerability coverage without expanding the proof surface past that gate.

File: internal/sourceprojection/snyk.go [hunk 8]
```diff
@@ -19,6 +19,42 @@ func snykVulnerabilitiesProjections(event *cerebrov1.EventEnvelope) ([]*ports.Pr
 	return snykGenericFindingProjections(event)
 }
 
+func snykOrgMembershipsProjections(event *cerebrov1.EventEnvelope) ([]*ports.ProjectedEntity, []*ports.ProjectedLink, error) {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceprojection/snyk.go:22]: Fixed in 0c3f11015. Snyk org/group membership projections now link members to the canonical plural inventory URNs (snyk_orgs/snyk_groups), with regression coverage for both org and group memberships.

File: internal/sourceprojection/snyk.go [hunk 9]
```diff
@@ -19,6 +19,42 @@ func snykVulnerabilitiesProjections(event *cerebrov1.EventEnvelope) ([]*ports.Pr
...
+func snykServiceAccountsProjections(event *cerebrov1.EventEnvelope) ([]*ports.ProjectedEntity, []*ports.ProjectedLink, error) {
+	return snykServiceAccountProjections(event)
+}
+
+func snykGroupServiceAccountsProjections(event *cerebrov1.EventEnvelope) ([]*ports.ProjectedEntity, []*ports.ProjectedLink, error) {
+	return snykServiceAccountProjections(event)
+}
+
+func snykAuditLogsProjections(event *cerebrov1.EventEnvelope) ([]*ports.ProjectedEntity, []*ports.ProjectedLink, error) {
+	return identityAuditProjections(event, identityProjectionProfile{Provider: "snyk"})
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceprojection/snyk.go]: Fixed in 0c3f11015. Snyk audit projections now map acted_on resources to canonical plural inventory URNs such as snyk_projects/snyk_groups/snyk_orgs, with updated projection tests.

File: sources/internal/snykapi/snykapi.go [hunk 10]
```diff
@@ -0,0 +1,652 @@
...
+	})
+}
+
+func snykAuditLogsFamily() jsonapi.Family {
+	family := snykOrgPagedFamily(jsonapi.Family{
+		Name:          familyAuditLogs,
+		Path:          "/orgs/{org_id}/audit_logs/search",
+		PathParams:    []string{orgIDConfig},
+		URNKind:       "snyk_audit_logs",
+		IDKeys:        []string{"created", "event", "content.user_id", "content.email"},
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sources/internal/snykapi/snykapi.go:327]: Fixed. Added a narrow jsonapi `IDTemplate` hook and wired Snyk org/group audit logs to use `${created}-${event}` for the visible record ID, exposed through `_record_id` for `source_event_id`. Also added audit-log `IdentityKeys` for event, actor, and resource scope so same-timestamp records no longer dedupe away before emission. Live Snyk coverage now asserts external_id/source_event_id and includes a two-event same-timestamp regression. Verified with `go test ./sources/internal/jsonapi ./sources/snyk ./sources/internal/snykapi ./internal/sourceprojection -count=1` and `make catalog-check detection-catalog-check connector-contract-check`.

File: internal/sourceprojection/snyk.go [hunk 11]
```diff
@@ -39,6 +75,211 @@ func snykGenericAssetProjections(event *cerebrov1.EventEnvelope) ([]*ports.Proje
...
+	}
+	attributes := event.GetAttributes()
+	assetID := firstNonEmpty(attributes["asset_id"], attributes["source_asset_id"])
+	relatedID := firstNonEmpty(attributes[relatedIDKey], attributes["resource_id"], event.GetId())
+	assetURN := projectionURN(tenantID, "snyk_assets", assetID)
+	relatedURN := projectionURN(tenantID, relatedURNKind, relatedID)
+	entities := map[string]*ports.ProjectedEntity{}
+	links := map[string]*ports.ProjectedLink{}
+	if assetID != "" {
+		addEntity(entities, &ports.ProjectedEntity{
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceprojection/snyk.go]: Fixed in 82ff26a35. Asset relationship projections no longer upsert a competing snyk.assets entity for the same URN, so inventory projections keep the runtime.* classification. The relationship still emits the associated_with link from the Snyk asset URN to the related project/target, and the projection test now asserts the conflicting asset entity is not created.

File: internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml [hunk 12]
```diff
@@ -341,25 +319,238 @@ entries:
...
+        - default_enabled: true
+          families: [assets, cloud_environments, cloud_resources]
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml]: Fixed in 4dc5743dd. The connector catalog now exposes all seven missing runtime families: collections, cloud_scans, group_memberships, group_service_accounts, group_audit_logs, asset_project_relationships, and asset_target_relationships. I added group_id/asset_id config fields, group access/audit and collections scopes, expanded cloud inventory, and validated with connector-contract-check, detection-catalog-check, focused Go tests, and structural/arch checks.

File: internal/findings/coverage_control_domain_refs.yaml [hunk 13]
```diff
@@ -161,6 +161,20 @@ control_domains:
...
+    - { framework_name: "NIST 800-53 r5", control_id: "AU-2" }
+    - { framework_name: "NIST 800-53 r5", control_id: "AU-6" }
+    - { framework_name: "NIST 800-53 r5", control_id: "SI-4" }
+    - { framework_name: "CIS Controls v8", control_id: "8" }
+    - { framework_name: "ISO 27001:2022", control_id: "A.8.15" }
+    - { framework_name: "ISO 27001:2022", control_id: "A.8.16" }
+    - { framework_name: "HIPAA", control_id: "164.312(b)" }
+    - { framework_name: "ISO 42001", control_id: "9.1" }
+    - { framework_name: "EU AI Act", control_id: "Art.12" }
+    - { framework_name: "EU AI Act", control_id: "Art.72" }
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/findings/coverage_control_domain_refs.yaml]: Fixed in 8495c8cec. Removed the duplicate `audit_logging` control-domain mapping and moved Snyk audit coverage to the existing `logging_monitoring` domain in both the connector catalog source and generated Snyk runtime catalog. The remaining `audit_logging` references are data-field names in policy queries, not control domains.

Validated with:
- `make detection-catalog-check catalog-check`
- `go test ./sources/snyk ./internal/sourceprojection -run 'TestSnyk' -count=1`\n- `git diff --check`

File: internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml [hunk 14]
```diff
@@ -341,25 +327,554 @@ entries:
...
+        - default_enabled: true
+          families: [collections]
+          id: collections
+          label: Collections
+        - default_enabled: true
+          families: [assets, cloud_environments, cloud_resources, cloud_scans, asset_project_relationships, asset_target_relationships]
+          id: cloud-inventory
+          label: Cloud inventory
+          needs_user_review: true
+          permission_note: Asset relationship families require asset_id or asset_ids runtime config.
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml:879]: Restored the persisted `asset-inventory` scope ID as a default-enabled scope over the original `projects`, `targets`, and `assets` families while keeping the newer application/cloud scopes. Verified with `make detection-catalog-check catalog-check`, `go test ./internal/connectorcatalog -count=1`, and the broader Snyk/sourceprojection package pass.

File: internal/sourceprojection/snyk.go [hunk 15]
```diff
@@ -39,6 +75,201 @@ func snykGenericAssetProjections(event *cerebrov1.EventEnvelope) ([]*ports.Proje
...
+		return "snyk_orgs"
+	case "group":
+		return "snyk_groups"
+	case "project":
+		return "snyk_projects"
+	case "target":
+		return "snyk_targets"
+	case "asset":
+		return "snyk_assets"
+	default:
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceprojection/snyk.go:222]: Fixed the audit resource mapping for collections, cloud environments, and cloud scans so `acted_on` links target the same plural inventory URN kinds emitted by the inventory projectors. Added table coverage for project, collection, cloud environment, and cloud scan audit resources. Verified with `go test ./internal/sourceprojection -run 'TestSnyk(AuditProjectionLinksActorToInventoryResource|RuntimeKindsAreExplicitDepthEvidence|AssetRelationshipProjectionLinksAssetToProjectAndTarget)' -count=1 -v` plus `go test ./internal/sourceprojection ./sources/snyk ./sources/internal/snykapi ./sources/internal/jsonapi ./internal/connectorcatalog -count=1`.

File: sources/internal/snykapi/snykapi.go [hunk 16]
```diff
@@ -0,0 +1,656 @@
...
+			"resource_id":     "project_id|org_id|group_id",
+			"resource_type":   "content.type|event",
+			"observed_at":     "created",
+		},
+		StaticAttributes: snykStaticAttributes("audit_logs", "audit_event", "snyk_audit_log"),
+	})
+	family.CursorParam = "cursor"
+	family.PageSizeParams = []string{"size"}
+	family.Config.IDTemplate = "${created}-${event}"
+	family.Config.IdentityKeys = []string{"event", "content.user_id", "content.email", "project_id", "org_id", "group_id"}
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sources/internal/snykapi/snykapi.go]: Fixed in cb0b17dad. Audit log external_id/source_event_id now include timestamp, event, actor fallback, and target fallback, so same-timestamp same-event records with different actors or targets do not collide. Updated the source tests to cover the same event type edge case directly.

File: internal/sourceprojection/snyk.go [hunk 17]
```diff
@@ -39,6 +75,207 @@ func snykGenericAssetProjections(event *cerebrov1.EventEnvelope) ([]*ports.Proje
...
+	case "asset":
+		return "snyk_assets"
+	case "collection":
+		return "snyk_collections"
+	case "cloud_environment", "environment":
+		return "snyk_cloud_environments"
+	case "cloud_scan", "scan":
+		return "snyk_cloud_scans"
+	default:
+		return "snyk_" + normalizeIdentifier(resourceType)
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceprojection/snyk.go]: Fixed in cb0b17dad. Snyk audit projection now returns no inventory URN kind for unmapped resource types, so event-name/resource fallback values do not create junk resource entities or acted_on links. Added a regression for group.membership.invite while keeping known inventory resource links covered.

File: sources/snyk/source_test.go [hunk 18]
```diff
@@ -229,6 +404,58 @@ func TestRuntimeUsesSnykRESTPathsAndVersionedPagination(t *testing.T) {
...
+		if got := attrs["external_id"]; got != want {
+			t.Fatalf("event %d external_id = %q, want %q", idx, got, want)
+		}
+		if got := attrs["source_event_id"]; got != want {
+			t.Fatalf("event %d source_event_id = %q, want %q", idx, got, want)
+		}
+	}
+	if pull.Events[0].Id == pull.Events[1].Id {
+		t.Fatalf("same-timestamp audit logs collapsed to duplicate event id %q", pull.Events[0].Id)
+	}
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sources/snyk/source_test.go:456]: Addressed in 6fdf49e05. Added TestGroupAuditLogsDoNotDedupeSameTimestamp alongside the org audit regression test, covering distinct IDs/source_event_id values for same-timestamp group audit events. Verified with go test ./sources/snyk ./sources/internal/snykapi ./sources/internal/jsonapi ./internal/sourceprojection ./internal/connectorcatalog -count=1 and make check-structural.

File: sources/internal/snykapi/snykapi.go [hunk 19]
```diff
@@ -0,0 +1,686 @@
...
+			"actor_email":     "content.email|content.user.email|content.actor.email",
+			"resource_id":     "project_id|org_id|group_id",
+			"resource_type":   "content.type|event",
+			"observed_at":     "created",
+		},
+		StaticAttributes: snykStaticAttributes("group_audit_logs", "audit_event", "snyk_audit_log"),
+	})
+	family.CursorParam = "cursor"
+	family.PageSizeParams = []string{"size"}
+	family.Config.IDTemplate = "${created}-${event}-" +
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sources/internal/snykapi/snykapi.go:554]: Addressed in 6fdf49e05. Group audit logs now mirror the org audit composite ID template and identity keys, and TestGroupAuditLogsDoNotDedupeSameTimestamp covers same-timestamp group audit events with distinct external_id/source_event_id values. Verified with go test ./sources/snyk ./sources/internal/snykapi ./sources/internal/jsonapi ./internal/sourceprojection ./internal/connectorcatalog -count=1 and make check-structural.

File: internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml [hunk 20]
```diff
@@ -341,25 +327,558 @@ entries:
...
-            - vulnerabilities
+          families: [collections]
+          id: collections
+          label: Collections
+        - default_enabled: true
+          families: [assets, cloud_environments, cloud_resources, cloud_scans, asset_project_relationships, asset_target_relationships]
+          id: cloud-inventory
+          label: Cloud inventory
+          needs_user_review: true
+          permission_note: Asset relationship families require asset_id or asset_ids runtime config.
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml:879]: Fixed in 3ce515d: scope_options now use disjoint family sets. application-inventory owns projects/targets, asset-inventory owns assets, collections remains standalone, and cloud-inventory owns cloud/relationship families. Verified with make connector-contract-check and make contracts-check.

File: internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml [hunk 21]
```diff
@@ -341,25 +327,558 @@ entries:
...
+            page_size: 100
+            page_size_param: size
+            type: cursor
+          path: /groups/${config.group_id}/audit_logs/search
+          projection:
+            fields:
+              actor_email: content.email|content.user.email|content.actor.email
+              actor_id: content.user_id|content.user.id|content.actor.id
+              event_type: event
+              group_id: ${config.group_id}
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/connectorcatalog/catalog/security-posture-vulnerability/snyk.yaml]: Fixed in c7ab981ad. `group_audit_logs` now declares `group_id: group_id|${config.group_id}` so connector catalog projection metadata matches the runtime family mapping and the sibling audit-log fallback shape. Verified with `go test ./internal/connectorcatalog -count=1`, `make connector-contract-check`, `make detection-catalog-check`, and `./scripts/leak_check.sh range origin/main...HEAD`.

File: internal/sourceprojection/snyk.go [hunk 22]
```diff
@@ -39,6 +75,210 @@ func snykGenericAssetProjections(event *cerebrov1.EventEnvelope) ([]*ports.Proje
...
+
+func snykAssetRelationshipProjections(event *cerebrov1.EventEnvelope, relatedIDKey string, relatedURNKind string, relatedType string) ([]*ports.ProjectedEntity, []*ports.ProjectedLink, error) {
+	tenantID, err := tenantID(event)
+	if err != nil {
+		return nil, nil, err
+	}
+	attributes := event.GetAttributes()
+	assetID := firstNonEmpty(attributes["asset_id"], attributes["source_asset_id"])
+	relatedID := firstNonEmpty(attributes[relatedIDKey], attributes["resource_id"], event.GetId())
+	assetURN := projectionURN(tenantID, "snyk_assets", assetID)
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceprojection/snyk.go:287]: Fixed in 865f37156. Snyk asset events now create a stable `snyk_assets:<asset_id>` alias entity and, when a runtime resource URN is also available or derivable, link that alias to the runtime resource with `represents`. Asset relationship and audit projections can now point at `snyk_assets:<asset_id>` without dangling against an entity that never exists. Added tests for the alias/runtime link and asset audit resource projection. Verified with `go test ./internal/sourceprojection ./sources/snyk -count=1`, `git diff --check`, and leak check.

File: internal/sourceprojection/snyk.go [hunk 23]
```diff
@@ -39,6 +81,210 @@ func snykGenericAssetProjections(event *cerebrov1.EventEnvelope) ([]*ports.Proje
...
+	case "collection":
+		return "snyk_collections"
+	case "cloud_environment", "environment":
+		return "snyk_cloud_environments"
+	case "cloud_scan", "scan":
+		return "snyk_cloud_scans"
+	default:
+		return ""
+	}
+}
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceprojection/snyk.go:225]: Fixed. Audit resource type mapping now covers `cloud_resource` with `snyk_cloud_resources` and keeps generic `resource` mapped to `snyk_assets`; the audit projection table test includes the cloud-resource case.

File: sources/internal/jsonapi/http.go [hunk 24]
```diff
@@ -762,16 +774,46 @@ func recordFromRaw(family Family, raw json.RawMessage) (record, error) {
...
+	if id == "" {
+		id = firstValueString(values, family.IDKeys...)
+	}
 	if id == "" {
 		if family.RequireID {
 			return record{}, fmt.Errorf("%s id is required", family.Name)
 		}
 		id = stableID(string(raw))
 	}
+	values["_record_id"] = id
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sources/internal/jsonapi/http.go:882]: Fixed in 253ad9b49. Detail-enriched records now strip _record_id before payload marshalling, with a regression assertion that the internal record id is not emitted.

File: internal/findings/public_detection_catalog.json [hunk 25]
```diff
@@ -280925,24 +280925,34 @@
...
+            {
+              "framework_name": "CIS Controls v8",
+              "control_id": "6"
+            },
             {
               "framework_name": "SOC 2",
-              "control_id": "CC6.1"
+              "control_id": "CC6.6"
             }
           ]
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/findings/public_detection_catalog.json]: Fixed in d97fd17d0. Partial audit-event coverage now requires a source match before it can attach to a policy detection, and partial audit preservation no longer evicts supported entity coverage under the cap. Regression tests cover both the capped PagerDuty users case and the non-capped Kubernetes authentication wording case.

File: internal/sourceprojection/snyk.go [hunk 26]
```diff
@@ -39,6 +81,212 @@ func snykGenericAssetProjections(event *cerebrov1.EventEnvelope) ([]*ports.Proje
...
+	case "cloud_environment", "environment":
+		return "snyk_cloud_environments"
+	case "cloud_resource":
+		return "snyk_cloud_resources"
+	case "cloud_scan", "scan":
+		return "snyk_cloud_scans"
+	default:
+		return ""
+	}
+}
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceprojection/snyk.go:225]: Fixed in c344f69ba. Audit resource resolution now maps `membership` resource types to the containing Snyk group or org scope, so membership audit events project an acted_on link to a real inventory entity. Added regression coverage for both group and org membership audit events while keeping compound unmapped event strings on the graceful-degradation path.

File: internal/sourceprojection/snyk.go [hunk 27]
```diff
@@ -39,6 +81,212 @@ func snykGenericAssetProjections(event *cerebrov1.EventEnvelope) ([]*ports.Proje
...
+		return "snyk_projects"
+	case "target":
+		return "snyk_targets"
+	case "asset", "resource":
+		return "snyk_assets"
+	case "collection":
+		return "snyk_collections"
+	case "cloud_environment", "environment":
+		return "snyk_cloud_environments"
+	case "cloud_resource":
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceprojection/snyk.go:218]: Fixed in 39040dffa. Snyk audit `cloud_resource` targets now resolve to the `snyk_assets` alias, which is the stable alias emitted by the cloud resource inventory projector. Added a regression test that projects a cloud resource inventory event and verifies the audit acted_on link points at the same asset alias rather than a disconnected `snyk_cloud_resources` stub.

File: sources/internal/snykapi/snykapi.go [hunk 28]
```diff
@@ -0,0 +1,686 @@
...
+func PathParamValues(cfg sourcecdk.Config, family string) (string, []string) {
+	switch strings.TrimSpace(family) {
+	case FamilyGroupMemberships, FamilyGroupSvcAccounts, FamilyGroupAuditLogs:
+		return GroupIDConfig, configListValues(cfg, "group_ids", GroupIDConfig)
+	case FamilyAssetProjects, FamilyAssetTargets:
+		return AssetIDConfig, configListValues(cfg, "asset_ids", AssetIDConfig)
+	default:
+		return "", nil
+	}
+}
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sources/internal/snykapi/snykapi.go:109]: Verified and covered in 000648837. Added Snyk adapter tests that read group-scoped families across comma-separated `group_ids` and asset-scoped families across comma-separated `asset_ids`, asserting both requested paths and emitted `group_id`/`asset_id` attributes. The lower-level `ReadPathParamValues` cursor behavior remains covered in `sources/internal/jsonapi/source_test.go`. Verified with focused Snyk/jsonapi/sourceprojection tests, scoped golangci-lint, policy-mapping-check, detection-catalog-check, and git diff --check.

File: internal/sourceprojection/snyk.go [hunk 29]
```diff
@@ -39,6 +81,229 @@ func snykGenericAssetProjections(event *cerebrov1.EventEnvelope) ([]*ports.Proje
...
+}
+
+func snykAuditResourceURNKind(event *cerebrov1.EventEnvelope, resourceType string) string {
+	if urnKind := snykInventoryURNKind(resourceType); urnKind != "" {
+		return urnKind
+	}
+	switch normalizeIdentifier(resourceType) {
+	case "membership", "group_membership", "org_membership":
+		attributes := event.GetAttributes()
+		if strings.TrimSpace(attributes["group_id"]) != "" || normalizeIdentifier(event.GetKind()) == "snyk_group_audit_logs" {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceprojection/snyk.go]: Fixed in c01ff9763. The fallback now compares the normalized event kind with the actual dot-form `snyk.group_audit_logs`, and `TestSnykAuditProjectionLinksMembershipEventsToScope` includes a case where `resource_type=membership` has no `group_id` but still links the acted-on resource to `snyk_groups`. Verified with focused Snyk/sourceprojection/jsonapi tests and scoped golangci-lint.
